src/crypto.rs

```diff
@@ -0,0 +1,92 @@
+use snafu::Snafu;
+
+use crate::xmlsec;
+use crate::xmlsec::XmlSecKey;
+use crate::xmlsec::XmlSecKeyFormat;
+use crate::xmlsec::XmlSecSignatureContext;
+use libxml::parser::Parser as XmlParser;
+
+
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    InvalidSignature,
+
+    #[snafu(display("xml sec Error: {}", error))]
+    XmlParseError {
+        error: libxml::parser::XmlParseError,
+    },
+
+    #[snafu(display("xml sec Error: {}", error))]
+    XmlSecError {
+        error: xmlsec::XmlSecError,
+    }
+}
+
+impl From<xmlsec::XmlSecError> for Error {
+    fn from(error: xmlsec::XmlSecError) -> Self {
+        Error::XmlSecError { error }
+    }
+}
+
+impl From<libxml::parser::XmlParseError> for Error {
+    fn from(error: libxml::parser::XmlParseError) -> Self {
+        Error::XmlParseError { error }
+    }
+}
+
+pub fn sign_xml<Bytes: AsRef<[u8]>>(xml: Bytes, private_key_der:&[u8]) -> Result<String, Error> {
+    let parser = XmlParser::default();
+    let document = parser.parse_string(xml)?;
+
+    let key = XmlSecKey::from_memory(private_key_der, XmlSecKeyFormat::Der)?;
+    let mut context = XmlSecSignatureContext::new()?;
+    context.insert_key(key);
+
+    context.sign_document(&document, Some("ID"))?;
+
+    Ok(document.to_string())
+}
+
+pub fn verify_signed_xml<Bytes: AsRef<[u8]>>(xml: Bytes, x509_cert_der: &[u8], id_attribute: Option<&str>) -> Result<(), Error> {
+    let parser = XmlParser::default();
+    let document = parser.parse_string(xml)?;
+
+    let key = XmlSecKey::from_memory(x509_cert_der, XmlSecKeyFormat::CertDer)?;
+    let mut context = XmlSecSignatureContext::new()?;
+    context.insert_key(key);
+
+    let valid = context.verify_document(&document, id_attribute)?;
+
+    if !valid {
+        return Err(Error::InvalidSignature);
+    }
+
+    Ok(())
+}
+
+// Util
+// strip out 76-width format and decode base64
+pub fn decode_x509_cert(x509_cert: &str) -> Result<Vec<u8>, base64::DecodeError> {
```

Comment:
I'd rather pull in https://crates.io/crates/line-wrap to support MIME line wrapping instead of a second base64 library. Thoughts?

Reply:
I agree in principle and also wanted to use line-wrap at first. I didn't like the interface to it though: it panics when you give it too small of a buffer: 

```
let mut data = vec![0; 82];
assert_eq!(2, line_wrap(&mut data, 80, 76, &crlf()));
```

Thoughts about removing the other base64 lib?